### PR TITLE
Port mosaic-chains to TypeScript

### LINKED ISSUE #18: Port to TypeScript
- [x] Port project to TypeScript
  - [x] Mosaic start
  - [x] Mosaic stop
  - [x] Mosaic attach
  - [x] Mosaic list
  - [x] Mosaic logs
  - [x] Documentation
  - [x] Tests

### PR DESCRIPTION
**Pull request by schemar**
The complexity of the project outgrew my Bash skills and was therefore
moved to TypeScript.

The command line interface has not changed at all. You still run the
same commands using `./mosaic`. However, `./mosaic` will now invoke the
TypeScript inside the `src` directory.

The scripts in `./src/bin` use the `commander` package to provide the
known mosaic commands.

A chain factory builds the two types of chains:

1. official chains (parity)
2. and utility chains (geth)

Fixes #18

### CODE REVIEW

File: src/bin/mosaic-attach.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  .arguments('<chain>')
+  .action((chainId) => {
+    const args = [
+      'run',
+      '--network', `${Chain.network}`,
+      '-it',
+      '--rm',
+      'ethereum/client-go',
+      'attach',
+      `http://${Chain.prefix}${chainId}:8545`
```
Comment (pgev) [inline src/bin/mosaic-attach.ts]: Can we document this part and explain why 8545 and also the magic http composition.

File: src/bin/mosaic-start.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env node
```
Comment (pgev) [inline src/bin/mosaic-start.ts]: I would propose not to have multiple chains as an input and stick with one.

File: .eslintrc.json [hunk 3]
```diff
@@ -0,0 +1,31 @@
+{
```
[the remark this answers is not in the document]
Comment (schemar) [reply .eslintrc.json]: > What about to use .eslintrc (without ".json") and align it. Currently openst ones are using .eslintrc and mosaic ones are using .eslintrc.json I believe.

The developer guidelines say `.eslintrc.json` right now 😉 https://github.com/OpenST/developer-guidelines/

File: README.md [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # ⛓ Mosaic Chains
 
-You need docker installed to run the chains.
+You need node and docker installed to run the chains.
 
 Mosaic will automatically identify if you want to run a utility chain (geth) or ethereum mainnet/testnet (parity).
```
[the remark this answers is not in the document]
Comment (schemar) [reply README.md]: I created #22
@pgev please finish #22 as I am not sure about the motivation.

File: README.md [hunk 5]
```diff
@@ -11,34 +11,26 @@ The default ports published with docker on the host are:
...
 
 When you start more than one chain, mosaic will increase all port numbers by 1 for each subsequent chain.
-If you already have containers or other services running at the default host ports for publishing, use the `-p`, `-r`, and/or `-w` options to start at a different port number.
+If you already have containers or other services running at the default host ports for publishing, use the `-port`, `-rpc-port`, and/or `-ws-port` options to start at a different port number.
 
 Stopping a container that was started with mosaic completely removes that container from the host.
 Only the content in the mounted data directory remains.
 
-⚠️ Nodes started with `mosaic` open *all* available APIs. If necessary, make sure your machine is secured or update `./mosaic` before running it.
+⚠️ Nodes started with `mosaic` open *all* available APIs. If necessary, make sure your machine is secured.
```
[the remark this answers is not in the document]
Comment (schemar) [reply README.md]: Created #23

File: README.md [hunk 6]
```diff
@@ -11,34 +11,26 @@ The default ports published with docker on the host are:
...
-Options:
-* `-d, --data-dir <path>`: a path to a directory where the chain data will be stored
-* `-p, --port <port>    `: the first port to use for forwarding from host to container
-* `-r, --rpc-port <port>`: the first RPC port to use for forwarding from host to container
-* `-w, --ws-port <port> `: the first WS port to use for forwarding from host to container
+Clone `git clone git@github.com:OpenST/mosaic-chains.git` and install `npm install`.
+Run `./mosaic` to get the help output.
 
 Examples:
 * Starts three containers to follow these chains:
```
[the remark this answers is not in the document]
Comment (schemar) [reply README.md]: They would each have a subdirectory in the shared `--data-dir`. The default is `~/.mosaic` (will add to README).

The port behavior is described correctly.

File: README.md [hunk 7]
```diff
@@ -47,10 +39,17 @@ Examples:
 * Follows the logs of a given chain:
   * `./mosaic logs ropsten`
 
+* Lists all running mosaic chain containers:
+  * `./mosaic list`
+
+## Tests
+
+Run the tests with `npm test`.
```
[the remark this answers is not in the document]
Comment (schemar) [reply README.md]: `npm test` is the default command to invoke tests on an npm project.

File: README.md [hunk 8]
```diff
@@ -47,10 +39,17 @@ Examples:
...
+
 ## Adding a new auxiliary chain
```
[the remark this answers is not in the document]
Comment (schemar) [reply README.md]: This is exactly what #12 is about 😊

File: package.json [hunk 9]
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "@openst/mosaic-chains",
+  "version": "0.1.0",
+  "description": "Mosaic chains is an executable to run and manage mosaic chains.",
+  "main": "index.js",
+  "scripts": {
+    "lint": "eslint src -c .eslintrc.json --ext ts",
```
[the remark this answers is not in the document]
Comment (schemar) [reply package.json]: Good point. You could bring it to https://github.com/OpenST/developer-guidelines and we could discuss it there 👍

File: package.json [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    "test": "mocha -r ts-node/register tests/**/*.test.ts tests/*.test.ts"
```
[the remark this answers is not in the document]
Comment (schemar) [reply package.json]: I don't want to use `transpile-only` in tests if they run in an acceptable time to not lose any benefits from TypeScript.

File: src/bin/mosaic-start.ts (hunk 2 above)
Comment (schemar) [reply src/bin/mosaic-start.ts]: Can you please create a new ticket? In my opinion, this is definitely out of the scope of this ticket.

File: mosaic [hunk 11]
```diff
@@ -1,211 +1,3 @@
...
-else
-  echo "Error: '$1' is not a valid action"
-  echo ""
-  help
-  exit 1
-fi
-
-exit 0
-
+TS_NODE_FILES=true; TS_NODE_TRANSPILE_ONLY=true; node -r ts-node/register ./src/bin/mosaic.ts $@
```
[the remark this answers is not in the document]
Comment (schemar) [reply mosaic]: Running `ts-node ...` does not work as it does not properly forward the output from the commander sub-commands.

**Comment by schemar** [comment]
@pgev I addressed all your comments, please have a look 😊

**Comment by benjaminbollen** [comment]
Lgtm

**Pull request merged**